File: QRID/QRID/Helpers/EventNotificationProtocol.swift
import Foundation

struct MeQREventCertificate: Codable, Sendable {
    let version: Int
    let eventID: String
    let channelID: String
    let eventName: String
    let publicKey: String
    let validFrom: Date
    let validUntil: Date
    let signature: String
}

struct MeQREventCheckInChallenge: Codable, Sendable {
    let eventID: String
    let challenge: String
    let issuedAt: Date
    let expiresAt: Date
    let signature: String
}

struct MeQREventNotification: Codable, Identifiable, Sendable {
    let id: String
    let eventID: String
    let title: String
    let body: String
    let publishedAt: Date
    let expiresAt: Date?
    let hop: Int
    let signature: String
}

enum MeQREventNotificationProtocol {
    static let protocolVersion = 1
    static let maxHopCount = 8
    static let maxNotificationBytes = 64 * 1024
    static func verifyCertificate(_ certificate: MeQREventCertificate) -> Bool { false }
    static func verifyCheckIn(_ challenge: MeQREventCheckInChallenge, certificate: MeQREventCertificate) -> Bool { false }
    static func verifyNotification(_ notification: MeQREventNotification, certificate: MeQREventCertificate) -> Bool { false }
}

File: Android/app/src/main/java/com/lucasli/meqr/EventNotificationProtocol.java
package com.lucasli.meqr;

final class EventNotificationProtocol {
    static final int VERSION = 1;
    static final int MAX_HOP_COUNT = 8;
    static final int MAX_NOTIFICATION_BYTES = 64 * 1024;

    static final class EventCertificate {
        final String eventId, channelId, eventName, publicKey, signature;
        final long validFrom, validUntil;
        EventCertificate(String eventId, String channelId, String eventName, String publicKey,
                         long validFrom, long validUntil, String signature) {
            this.eventId = eventId; this.channelId = channelId; this.eventName = eventName;
            this.publicKey = publicKey; this.validFrom = validFrom; this.validUntil = validUntil;
            this.signature = signature;
        }
    }

    static final class CheckInChallenge {
        final String eventId, challenge, signature;
        final long issuedAt, expiresAt;
        CheckInChallenge(String eventId, String challenge, long issuedAt, long expiresAt, String signature) {
            this.eventId = eventId; this.challenge = challenge; this.issuedAt = issuedAt;
            this.expiresAt = expiresAt; this.signature = signature;
        }
    }

    private EventNotificationProtocol() { }
}
